Core/src/com/infiniteautomation/mango/spring/service/ConfigurationTemplateService.java

```diff
@@ -0,0 +1,155 @@
+            PermissionException {
+        MustacheFactory mf = new DefaultMustacheFactory();
+        List<Map<String, Object>> sites = generateConfig(fileStore, filePath);
+
+        ClassPathResource cp = new ClassPathResource(template);
+        Mustache m = mf.compile(cp.getPath());
+        StringWriter writer = new StringWriter();
+        m.execute(writer, sites).flush();
+        System.out.println("Testing result" + writer.toString());
+        return writer.toString();
+    }
+
+    public static List<Map<String, Object>> groupByKey(List<Map<String, Object>> array, Map<String, String> key) {
+        Map<Object, Map<String, Object>> map = new HashMap<>();
+        for (Map<String, Object> item : array) {
+            Object keyValue = item.get(key.get("groupKey"));
+            if (!map.containsKey(keyValue)) {
+                Map<String, Object> group = new HashMap<>();
+                group.put(key.get("groupKey"), keyValue);
+                group.put(key.get("childrenKey"), new ArrayList<Map<String, Object>>());
+                map.put(keyValue, group);
+            }
+            ((List)map.get(keyValue).get(key.get("childrenKey"))).add(item);
+        }
+        List<Map<String, Object>> groups = new ArrayList<>(map.values());
+
+        // find common properties from the children and copy them to the group
+        for (Map<String, Object> group : groups) {
+            Map<String, Object> common = findCommonProperties((List<Map<String, Object>>) group.get(key.get("childrenKey")));
+            group.putAll(common);
+        }
+
+        return groups;
+    }
+
+    public static List<Map<String, Object>> groupByKeys(List<Map<String, Object>> data, List<Map<String, String>> keys) {
+        if (keys.isEmpty()) {
+            return data;
+        }
+        Map<String, String> firstKey = keys.get(0);
+        List<Map<String, String>> nextKeys = keys.subList(1, keys.size());
+        List<Map<String, Object>> groups = groupByKey(data, firstKey);
+        for (Map<String, Object> group : groups) {
+            group.put(firstKey.get("childrenKey"), groupByKeys((List<Map<String, Object>>) group.get(firstKey.get("childrenKey")), nextKeys));
+        }
+        return groups;
+    }
+
+    public static Map<String, Object> findCommonProperties(List<Map<String, Object>> data) {
+        if (data.isEmpty()) {
+            return new HashMap<>();
+        }
+        Map<String, Object> common = new HashMap<>(data.get(0));
+        for (Map<String, Object> item : data) {
+            for (Map.Entry<String, Object> entry : new HashMap<>(common).entrySet()) {
+                String k = entry.getKey();
+                Object v = entry.getValue();
+                if (!item.get(k).equals(v)) {
+                    common.remove(k);
+                }
+            }
+        }
+        return common;
+    }
+
+    private List<Map<String, Object>> readCSV(CSVReader reader) throws IOException {
```

Comment:
Any specific reason these are public and static?

Reply:
none, changed to private methods.